### PR TITLE
OF-1309 Authenticate based on domain pairs

### PR DESCRIPTION
**Pull request by dwd**
Openfire considers that if a session has authenticated as A, and has
authenticated to B, then the session is already authorized to send stanzas
from A to B. This is not always the case.

This patch stores domain pairs (encapsulated as a DomainPair object) for a
session, and always authenticates if it has previously not requested
authorization for this domain pair.

This should be sufficient to fix OF-1309, but testing is tricky.

### CODE REVIEW

**Comment by dwd** [comment]
Haven't done any testing beyond "Does it still build?", but I could use the reviews...

File: src/java/org/jivesoftware/openfire/session/DomainPair.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+public class DomainPair {
+    private final String local;
+    private final String remote;
+
+    public DomainPair(String local, String remote) {
+        this.local = local;
+        this.remote = remote;
+    }
+
+    public int hashCode() {
```
Comment (guusdk) [inline src/java/org/jivesoftware/openfire/session/DomainPair.java]: Even though it is *probably* alright, the implementation of `hashCode` and `equals` aught to be consistent. By far the easiest way to guarantee this is to let your IDE generate these methods for you. In Intellij: [ALT-INSERT (select "generate equals() and hashCode()")](https://www.jetbrains.com/help/idea/generate-equals-and-hashcode-wizard.html). Generated code also lets you stop worry about null references and the like.

This is of particular importance here, as you're storing instances of `DomainPair` in a HashSet (in `LocalOutgoingServerSession`)

File: src/java/org/jivesoftware/openfire/session/LocalOutgoingServerSession.java [hunk 2]
```diff
@@ -698,4 +700,13 @@ public String getAvailableStreamFeatures() {
...
     }
+
+    public void addOutgoingDomainPair(String localDomain, String remoteDomain) {
+        outgoingDomainPairs.add(new DomainPair(localDomain, remoteDomain));
+    }
+
+    @Override
+    public boolean checkOutgoingDomainPair(String localDomain, String remoteDomain) {
+         return outgoingDomainPairs.contains(new DomainPair(localDomain, remoteDomain));
+    }
```
Comment (guusdk) [inline src/java/org/jivesoftware/openfire/session/LocalOutgoingServerSession.java]: Is there any point - or is it even safe - to let `#getAuthenticatedDomains()` exist? Isn't usage of that method inherently unsafe, in context of the issue that we're trying to fix. If so, `#addAuthenticatedDomain` can probably also go.

File: src/java/org/jivesoftware/openfire/session/OutgoingServerSession.java [hunk 3]
```diff
@@ -79,4 +79,6 @@
      * @return True if the subdomain was authenticated by the remote server.
      */
     boolean authenticateSubdomain(String domain, String hostname);
+
```
Comment (guusdk) [inline src/java/org/jivesoftware/openfire/session/OutgoingServerSession.java]: Tsk. All other definitions in this interface definition got nice documentation. :)

File: src/java/org/jivesoftware/openfire/session/DomainPair.java (hunk 1 above)
Comment (dwd) [reply src/java/org/jivesoftware/openfire/session/DomainPair.java]: Done.

File: src/java/org/jivesoftware/openfire/session/LocalOutgoingServerSession.java (hunk 2 above)
Comment (dwd) [reply src/java/org/jivesoftware/openfire/session/LocalOutgoingServerSession.java]: Yes, you're right - I've removed both getAuthenticatedDomains and getHostnames and so on, and replaced them purely with DomainPairs throughout.

File: src/java/org/jivesoftware/openfire/session/OutgoingServerSession.java (hunk 3 above)
Comment (dwd) [reply src/java/org/jivesoftware/openfire/session/OutgoingServerSession.java]: I have written nice documentation now.

**Comment by dwd** [comment]
Tested this, now can't find a behaviour difference for simple cases. I haven't figured out how to consistently trigger OF-1309 yet, but I'm confident it won't affect this design.

**Pull request merged**